### ISSUE TITLE
Inquiries on SpliceBERT implementation

### ISSUE THREAD

**Issue by qxdu** [chenkenbio/SpliceBERT#3]
Hi there,

I have downloaded your model `SpliceBERT.1024nt` from Zenodo and tried to retrain it. However, I encountered the following issues:

(1) I am curious about the version of your environment settings. I received the following warning when I loaded the provided `SpliceBERT.1024nt` model:

```
BertForMaskedLM has generative capabilities, as `prepare_inputs_for_generation` is explicitly overwritten. However, it doesn't directly inherit from `GenerationMixin`. From 👉v4.50👈 onwards, `PreTrainedModel` will NOT inherit from `GenerationMixin`, and this model will lose the ability to call `generate` and other related functions
```

And the output after the following codes are confusing:
```
# load tokenizer
tokenizer = AutoTokenizer.from_pretrained(SPLICEBERT_PATH)

# prepare input sequence
seq = "ACGUACGuacguaCGu"  ## WARNING: this is just a demo. SpliceBERT may not work on sequences shorter than 64nt as it was trained on sequences of 64-1024nt in length
seq = ' '.join(list(seq.upper().replace("U", "T"))) # U -> T and add whitespace
input_ids = tokenizer.encode(seq) # N -> 5, A -> 6, C -> 7, G -> 8, T(U) -> 9. NOTE: a [CLS] and a [SEP] token will be added to the start and the end of seq
input_ids = torch.as_tensor(input_ids) # convert python list to Tensor
input_ids = input_ids.unsqueeze(0) # add batch dimension, shape: (batch_size, sequence_length)

print(input_ids.shape)

# use huggerface's official API to use SpliceBERT
# get nucleotide embeddings (hidden states)
model = AutoModel.from_pretrained(SPLICEBERT_PATH) # load model
last_hidden_state = model(input_ids).last_hidden_state # get hidden states from last layer
hiddens_states = model(input_ids, output_hidden_states=True).hidden_states # hidden states from the embedding layer (nn.Embedding) and the 6 transformer encoder layers

# get nucleotide type logits in masked language modeling
model = AutoModelForMaskedLM.from_pretrained(SPLICEBERT_PATH) # load model
logits = model(input_ids).logits # shape: (batch_size, sequence_length, vocab_size)

# finetuning SpliceBERT for token classification tasks
model = AutoModelForTokenClassification.from_pretrained(SPLICEBERT_PATH, num_labels=3) # assume the class number is 3, shape: (batch_size, sequence_length, num_labels)
logits = model(input_ids).logits
print(logits.shape)

# finetuning SpliceBERT for sequence classification tasks
model = AutoModelForSequenceClassification.from_pretrained(SPLICEBERT_PATH, num_labels=3) # assume the class number is 3, shape: (batch_size, sequence_length, num_labels)
logits = model(input_ids).logits
print(logits)

tensor([[ 0.0757, -0.0088, -0.0167]], grad_fn=<AddmmBackward0>)
```

The values do not add up to 1. Is it a normal phenomenon, as you mentioned, that "SpliceBERT may not work on sequences shorter than 64nt as it was trained on sequences of 64-1024nt in length"?

(2) The output of your model `SpliceBERT.1024nt` appears to be a 3-channel output. I am unsure of the specific meaning of each channel and which figure in the main article it corresponds to.

(3) The output channels in your [examples](https://github.com/chenkenbio/SpliceBERT/blob/main/examples/05-bp-prediction/finetune_for_bp_prediction.py) seem to be 2, and in your paper, you describe “binary classification.” I am confused about the difference between the 2-output and 3-output channels and which model I should use.

Thank you in advance for your consideration and support.

**Comment by chenkenbio** [comment]
Hi, thanks for reaching out. 
(1.1) I used `transformers (4.24.0)` in this study (as specified in the readme). This issue was caused by the updates in the Huggingface transformer package, while I think it will not affect the outputs (except for generative tasks).
(1.2) logits should be normalized (usually using softmax) if you want to obtain probabilities.
(2 & 3) The `num_labels` argument in `AutoModelForTokenClassification.from_pretrained` should be adjusted for different tasks. For binary classification, you can set `num_labels` to $1$ (loss function should be binary cross entropy loss) or $2$ (loss function should be cross entropy loss). For $N$-class classification, you can set `num_labels` to $N$ accordingly. The last linear layer in the model will map the embedding of each token or sequence to an $N$-dimensional logits vector. (see source code: https://github.com/huggingface/transformers/blob/a6256ec0982fee2c57cc41237bff7e64ed4dcda9/src/transformers/models/bert/modeling_bert.py#L1827). 

![image](https://github.com/user-attachments/assets/42875c2f-2887-490b-b793-f9167049742b)

**Comment by chenkenbio** [comment]
Yeah. The model weights were obtained through masked language modeling in a self-supervised manner. Fine-tuning is needed for specific supervised tasks.
